difference-in-days xpath date functions fail in CEST timezone

Unit tests should pass regardless of timezone. Running `npm run unit-webapp` on my dev machine in Germany (CEST) I get...

```
  2 failing

  1) medic-xpath-extensions
       #difference-in-days
         should handle cases where end date is before start date and return negative number:

      AssertionError: expected -211 to equal -212
      + expected - actual

      --211
      +-212
      
      at Context.<anonymous> (webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js:179:14)
      at processImmediate (node:internal/timers:466:21)

  2) medic-xpath-extensions
       #difference-in-days
         should return the difference when valid inputs are provided:

      AssertionError: expected 211 to deeply equal 212
      + expected - actual

      -211
      +212
      
      at Context.<anonymous> (webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js:189:14)
      at processImmediate (node:internal/timers:466:21)

```

--- a/package.json
+++ b/package.json
@@ -40,7 +40,7 @@
     "unit-api": "UNIT_TEST_ENV=1 nyc --nycrcPath='api/nyc.config.js' mocha 'api/tests/mocha/**/*.js'",
     "unit-sentinel": "UNIT_TEST_ENV=1 nyc --nycrcPath='sentinel/nyc.config.js' mocha 'sentinel/tests/**/*.js'",
     "unit-shared-lib": "UNIT_TEST_ENV=1 npm test --workspaces --if-present",
-    "unit-webapp": "UNIT_TEST_ENV=1 mocha 'webapp/tests/mocha/**/*.spec.js' && cd webapp && npm run unit -- --watch=false && npm run unit:cht-form -- --watch=false",
+    "unit-webapp": "cd webapp && npm run unit:mocha:tz && npm run unit -- --watch=false && npm run unit:cht-form -- --watch=false",
     "unit-webapp-continuous": "cd webapp && npm run unit -- --watch=true",
     "unit-nginx": "cd nginx/tests && make test",
     "unit-haproxy": "cd haproxy/tests && make test",

--- a/webapp/package.json
+++ b/webapp/package.json
@@ -14,8 +14,10 @@
   },
   "scripts": {
     "postinstall": "patch-package",
-    "unit": "UNIT_TEST_ENV=1 ng test webapp",
+    "unit:mocha": "UNIT_TEST_ENV=1 mocha 'tests/mocha/**/*.spec.js'",
+    "unit:mocha:tz": "TZ=Canada/Pacific npm run unit:mocha && TZ=Africa/Monrovia npm run unit:mocha && TZ=Pacific/Auckland npm run unit:mocha",
     "unit:cht-form": "ng test cht-form",
+    "unit": "UNIT_TEST_ENV=1 ng test webapp",
     "build": "ng build",
     "build:cht-form": "ng build cht-form",
     "compile": "ngc"

--- a/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
+++ b/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
@@ -12,6 +12,9 @@ describe('medic-xpath-extensions', function () {
   };
 
   describe('getTimezoneOffsetAsTime()', function () {
+    const getTimeZoneOffset = Date.prototype.getTimezoneOffset;
+    afterEach(() => Date.prototype.getTimezoneOffset = getTimeZoneOffset);
+
     it('returns the time zone offset in hours when given a time zone difference in minutes', function () {
       Date.prototype.getTimezoneOffset = () => -60;
       assert.equal(medicXpathExtensions.getTimezoneOffsetAsTime(new Date()), '+01:00');
@@ -24,6 +27,9 @@ describe('medic-xpath-extensions', function () {
   });
 
   describe('toISOLocalString()', function () {
+    const getTimeZoneOffset = Date.prototype.getTimezoneOffset;
+    afterEach(() => Date.prototype.getTimezoneOffset = getTimeZoneOffset);
+
     it('returns the ISO local string consistent with the time zone offset', function () {
       const date = new Date('August 19, 1975 23:15:30 GMT+07:00');
       Date.prototype.getTimezoneOffset = () => -60;
@@ -93,7 +99,7 @@ describe('medic-xpath-extensions', function () {
       ['2015-10-01T11:11:11.111', '2014-10-01T11:11:11.111', -12,],
       ['2014-10-01T00:00:00.000', '2015-10-01T00:00:00.000', 12,],
       ['August 19, 1975 00:00:00 GMT', 'August 18, 1976 23:15:30 GMT+07:00', 11],
-      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2300', 0],
+      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 21:00:00 GMT+2300', 0],
       ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2200', 1]
     ].forEach(function (example) {
       const d1 = { t: 'str', v: example[0] };
@@ -106,8 +112,8 @@ describe('medic-xpath-extensions', function () {
     });
 
     it('should report difference between date objects', function () {
-      const d1 = { t: 'date', v: new Date('2015-09-22') };
-      const d2 = { t: 'date', v: new Date('2014-09-22') };
+      const d1 = { t: 'date', v: moment('2015-09-22').toDate() };
+      const d2 = { t: 'date', v: moment('2014-09-22').toDate() };
 
       assert.equal(diffInMonths(d1, d2).v, -12);
     });
@@ -199,14 +205,18 @@ describe('medic-xpath-extensions', function () {
 
   describe('#to-bikram-sambat()', () => {
     [
-      [{ t: 'str', v: '2015-9-2' }, '१६ भदौ २०७२'],
-      [{ t: 'str', v: '1999-12-12' }, '२६ मंसिर २०५६'],
-      [{ t: 'date', v: new Date('2015-10-01T00:00:00.000') }, '१४ असोज २०७२'],
-      [{ t: 'num', v: 11111 }, '२१ जेठ २०५७'],
-      [{ t: 'arr', v: [{ textContent: '2014-09-22' }] }, '६ असोज २०७१'],
-    ].forEach(([date, expected]) => {
+      [{ t: 'str', v: '2015-9-2' }, '१६ भदौ २०७२', '१५ भदौ २०७२'],
+      [{ t: 'str', v: '1999-12-12' }, '२६ मंसिर २०५६', '२५ मंसिर २०५६'],
+      [{ t: 'date', v: new Date('2015-10-01T00:00:00.000') }, '१४ असोज २०७२', '१३ असोज २०७२'],
+      [{ t: 'num', v: 11111 }, '२१ जेठ २०५७', '२० जेठ २०५७'],
+      [{ t: 'arr', v: [{ textContent: '2014-09-22' }] }, '६ असोज २०७१', '५ असोज २०७१'],
+    ].forEach(([date, expected, offsetExpected]) => {
       it(`should format the ${date.t} [${JSON.stringify(date.v)}] according to the Bikram Sambat calendar`, () => {
-        assert.equal(func['to-bikram-sambat'](date).v, expected);
+        // The resolved Bikram Sambat date for input that does not have a time+TZ value will be different in
+        // environments where the timezone offset is 12 hours or greater (e.g. `Pacific/Auckland`)
+        const offsetHrs = new Date().getTimezoneOffset() / 60;
+        const tzDateOffset = offsetHrs <= -12;
+        assert.equal(func['to-bikram-sambat'](date).v, tzDateOffset ? offsetExpected : expected);
       });
     });
 
@@ -235,7 +245,7 @@ describe('medic-xpath-extensions', function () {
     [
       [
         [{ t: 'num', v: 1 }],
-        { t: 'date', v: new Date('2015-10-01') },
+        { t: 'date', v: moment('2015-10-01').toDate() },
         '2016-10-01'
       ],
       [
@@ -265,7 +275,7 @@ describe('medic-xpath-extensions', function () {
       ],
       [
         [{ t: 'num', v: -1 }],
-        { t: 'date', v: new Date('2015-10-01') },
+        { t: 'date', v: moment('2015-10-01').toDate() },
         '2014-10-01'
       ],
       [
@@ -310,28 +320,28 @@ describe('medic-xpath-extensions', function () {
       ],
       [
         [{ t: 'arr', v: [{ textContent: '1' }] }],
-        { t: 'date', v: new Date('2015-10-01') },
+        { t: 'date', v: moment('2015-10-01').toDate() },
         '2016-10-01'
       ],
       [
         [{ t: 'num', v: 0 }, { t: 'num', v: 0 }, { t: 'num', v: 0 }, { t: 'num', v: 0 }, { t: 'num', v: 0 }],
-        { t: 'date', v: new Date('2015-10-01') },
+        { t: 'date', v: moment('2015-10-01').toDate() },
         '2015-10-01'
       ],
       [
         [],
-        { t: 'date', v: new Date('2015-10-01') },
+        { t: 'date', v: moment('2015-10-01').toDate() },
         '2015-10-01'
       ]
     ].forEach(([[year, month, day, hour, minute], date, expected]) => {
       const valuesDisplay = `[${display(year)},${display(month)},${display(day)},${display(hour)},${display(minute)}]`;
       it(`should add ${valuesDisplay} to ${date.t} [${JSON.stringify(date.v)}]`, () => {
-        assert.deepEqual(func['add-date'](date, year, month, day, hour, minute).v, new Date(expected));
+        assert.deepEqual(func['add-date'](date, year, month, day, hour, minute).v, moment(expected).toDate());
       });
     });
 
     it(`should throw an error when providing too many number parameters`, () => {
-      const date = { t: 'date', v: new Date('2015-10-01') };
+      const date = { t: 'date', v: moment('2015-10-01').toDate() };
       assert.throws(() => func['add-date'](date, 1, 2, 3, 4, 5, 6), 'Too many arguments.');
     });
 
@@ -339,14 +349,14 @@ describe('medic-xpath-extensions', function () {
       { t: 'str', v: '' },
       { t: 'bool', v: 'true' },
       { t: 'str', v: 'invalid' },
-      { t: 'date', v: new Date('2015-10-01') },
+      { t: 'date', v: moment('2015-10-01').toDate() },
       { t: 'num', v: 'one' },
       { t: 'num', v: 'NaN' },
       { t: 'num', v: NaN },
       { t: 'arr', v: [{ textContent: '2015-10-01' }] }
     ].forEach(year => {
       it(`should not modify the date when given ${year.t} [${JSON.stringify(year.v)}] as a number parameter`, () => {
-        const date = { t: 'date', v: new Date('2015-10-01') };
+        const date = { t: 'date', v: moment('2015-10-01').toDate() };
         assert.deepEqual(func['add-date'](date, year).v, date.v);
       });
     });

--- a/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
+++ b/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
@@ -93,7 +93,7 @@ describe('medic-xpath-extensions', function () {
       ['2015-10-01T11:11:11.111', '2014-10-01T11:11:11.111', -12,],
       ['2014-10-01T00:00:00.000', '2015-10-01T00:00:00.000', 12,],
       ['August 19, 1975 00:00:00 GMT', 'August 18, 1976 23:15:30 GMT+07:00', 11],
-      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 21:59:59 GMT+2300', 0],
+      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2300', 0],
       ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2200', 1]
     ].forEach(function (example) {
       const d1 = { t: 'str', v: example[0] };
@@ -175,7 +175,7 @@ describe('medic-xpath-extensions', function () {
     const diffInDays = func['cht:difference-in-days'];
 
     it('should handle cases where end date is before start date and return negative number', function () {
-      const result = diffInDays(wrapDate('2023-08-01 UTC'), wrapDate('2023-01-01 UTC'));
+      const result = diffInDays(wrapDate('2023-08-01'), wrapDate('2023-01-01'));
       assert.equal(result.v, -212);
     });
 
@@ -185,7 +185,7 @@ describe('medic-xpath-extensions', function () {
     });
 
     it('should return the difference when valid inputs are provided', function () {
-      const result = diffInDays(wrapDate('2023-01-01 UTC'), wrapDate('2023-08-01 UTC'));
+      const result = diffInDays(wrapDate('2023-01-01'), wrapDate('2023-08-01'));
       assert.deepStrictEqual(result.v, 212);
     });
 

--- a/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
+++ b/webapp/tests/mocha/unit/enketo/medic-xpath-extensions.spec.js
@@ -93,7 +93,7 @@ describe('medic-xpath-extensions', function () {
       ['2015-10-01T11:11:11.111', '2014-10-01T11:11:11.111', -12,],
       ['2014-10-01T00:00:00.000', '2015-10-01T00:00:00.000', 12,],
       ['August 19, 1975 00:00:00 GMT', 'August 18, 1976 23:15:30 GMT+07:00', 11],
-      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2300', 0],
+      ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 21:59:59 GMT+2300', 0],
       ['Sun Sep 25 2005 1:00:00 GMT+0100', 'Sun Oct 25 2005 22:00:00 GMT+2200', 1]
     ].forEach(function (example) {
       const d1 = { t: 'str', v: example[0] };
@@ -175,7 +175,7 @@ describe('medic-xpath-extensions', function () {
     const diffInDays = func['cht:difference-in-days'];
 
     it('should handle cases where end date is before start date and return negative number', function () {
-      const result = diffInDays(wrapDate('2023-08-01'), wrapDate('2023-01-01'));
+      const result = diffInDays(wrapDate('2023-08-01 UTC'), wrapDate('2023-01-01 UTC'));
       assert.equal(result.v, -212);
     });
 
@@ -185,7 +185,7 @@ describe('medic-xpath-extensions', function () {
     });
 
     it('should return the difference when valid inputs are provided', function () {
-      const result = diffInDays(wrapDate('2023-01-01'), wrapDate('2023-08-01'));
+      const result = diffInDays(wrapDate('2023-01-01 UTC'), wrapDate('2023-08-01 UTC'));
       assert.deepStrictEqual(result.v, 212);
     });
 

--- a/webapp/tests/mocha/unit/validate_doc_update.spec.js
+++ b/webapp/tests/mocha/unit/validate_doc_update.spec.js
@@ -38,12 +38,12 @@ describe('validate doc update', () => {
   };
 
   const clientValidateDocUpdate = function() {
-    const fn = fs.readFileSync('./ddocs/medic-db/medic-client/validate_doc_update.js');
+    const fn = fs.readFileSync(`${__dirname}/../../../../ddocs/medic-db/medic-client/validate_doc_update.js`);
     eval('log = console.log; (' + fn + ')').apply(null, arguments);
   };
 
   const serverValidateDocUpdate = function() {
-    const fn = fs.readFileSync('./ddocs/medic-db/medic/validate_doc_update.js');
+    const fn = fs.readFileSync(`${__dirname}/../../../../ddocs/medic-db/medic/validate_doc_update.js`);
     eval('log = console.log; (' + fn + ')').apply(null, arguments);
   };
 

